docs/content-flow.md

```diff
@@ -0,0 +1,415 @@
+containerd.io/gc.ref.content.2=sha256:f2edbd6a658e04d559c1bec36d838006bbdcb39d8fb9033ed43d2014ac497774
+containerd.io/gc.ref.content.3=sha256:66960bede47c1a193710cf8bfa7bf5f50bc46374260923df1db1c423b52153ac
+containerd.io/gc.ref.content.4=sha256:79dc0b596c9027416a627a6237bd080ac9d87f92b60f1ce145c566632839bce7
+containerd.io/gc.ref.content.5=sha256:de36df38e0b6c0e7f29913c68884a0323207c07cd7c1eba71d5618f525ac2ba6
+containerd.io/gc.ref.content.6=sha256:602cd484ff92015489f7b9cf9cbd77ac392997374b1cc42937773f5bac1ff43b
+```
+
+These are precisely those children of the manifest - the config and layers - that are stored in our content store.
+
+##### Index Labels
+
+The labels on the index also begin with `containerd.io/gc.ref`, ndicating that they are used to control
+garbage collection. An index has several "children", i.e. the manifests, one for each platform, as discussed above.
+We want to ensure that as long as the index remains around, the children do not get garbage collected.
+Thus, we have labels referencing each child, `containerd.io/gc.ref.content.<index>`.
+
+In our example, the index is `sha256:1d0b903e3770c2c3c79961b73a53e963f4fd4b2674c2c4911472e8a054cb5728`, and the labels are as follows:
+
+```
+containerd.io/gc.ref.content.0=sha256:a5aae2581826d13e906ff5c961d4c2817a9b96c334fd97b072d976990384156a
+containerd.io/gc.ref.content.1=sha256:4ff8940144391ecd5e1632d0c427d95f4a8d2bb4a72b7e3898733352350d9ab3
+containerd.io/gc.ref.content.2=sha256:ce541c3e2570b5a05d40e7fc01f87fc1222a701c81f95e7e6f2ef6df1c6e25e7
+containerd.io/gc.ref.content.3=sha256:535ee258100feeeb525d4793c16c7e58147c105231d7d05ffc9c84b56750f233
+containerd.io/gc.ref.content.4=sha256:0f3b047f2789547c58634ce88d71c7856999b2afc8b859b7adb5657043984b26
+containerd.io/gc.ref.content.5=sha256:bfc45f499a9393aef091057f3d067ff7129ae9fb30d9f31054bafe96ca30b8d6
+containerd.io/gc.ref.content.6=sha256:3198e1f1707d977939154a57918d360a172c575bddeac875cb26ca6f4d30dc1c
+containerd.io/gc.ref.content.7=sha256:24a15cc9366e1557db079a987e63b98a5abf4dee4356a096442f53ddc8b9c7e9
+```
+
+Notice that there are 8 children to the index, but all of them are for platforms other than ours, `linux/amd64`,
+and thus only one of them, `sha256:a5aae2581826d13e906ff5c961d4c2817a9b96c334fd97b072d976990384156a` actually is
+in our content store. That doesn't hurt; it just means that the others will not be garbage collected either. Since
+they aren't there, they won't be removed.
+
+### Snapshots
+
+The content in the content store is immutable, but also in formats that often are unusable. For example,
+most container layers are in a tar-gzip format. One cannot simply mount a tar-gzip file. Even if one could,
+we want to leave our immutable content not only unchanged, but unchangeable, even by accident, i.e. immutable.
+
+In order to use it, we create snapshots of the content.
+
+Each layer in an image is converted into a committed snapshot. The snapshotter itself knows how to extract data
```

Comment:
Nit here: the snapshotter doesn't know how to extract the data, this is done by the diff applier. The snapshotter is only concerned with providing a merged filesystem.

Reply:
Now we are into details I don't understand well enough. Do you mind explaining a little further, so I can capture it? I would put in the language you have here, but I never mentioned the diff applier before, so I would need to have some explanation of what it is.

Reply:
Can you give some extra insight here @cpuguy83 so I can fix it, and we can get it merged in?

Reply:
The snapshotter only cares about the storage aspect, as such it is, generally speaking, relegated to preparing mounts.
It does not understand layer diffs, whiteouts, etc.
This is up to the diff applier: https://godoc.org/github.com/containerd/containerd/diff#Applier (this is for the go client).
containerd provides a diff service to satisfy this interface.
The diff service is responsible for calculating layer diffs as well as applying layer diffs to a mount.

Reply:
Reading through this [from here](https://github.com/containerd/containerd/blob/master/rootfs/apply.go#L114-L179), it looks like the process is the following for each layer:

1. Create a new snapshot from the parent [here](https://github.com/containerd/containerd/blob/master/rootfs/apply.go#L129). This gives us a clean, new, active snapshot with nothing different from the parent.
1. Use the diff applier ("differ"?) to apply the layer (i.e. its diffs) to this new active snapshot [here](https://github.com/containerd/containerd/blob/master/rootfs/apply.go#L163)
1. Commit the new snapshot [here](https://github.com/containerd/containerd/blob/master/rootfs/apply.go#L173)

Repeat for each layer until done.

Snapshotter really is just "make a new point-in-time copy with unique mounts" (i.e. snapshot, hence the name) from a parent. The diff applier uses the layer to make changes to the new (active) snapshot, and then we commit it, locking it in place.

Is that about right?

Reply:
Take a look at the latest comment please @cpuguy83 ?

Reply:
SGTM

Reply:
OK, will update the PR and @ mention you.

Reply:
Updated @cpuguy83 ; this work?